dds/DCPS/InternalDataReader.h

```diff
@@ -184,62 +192,257 @@ class InternalDataReader : public InternalEntity {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+
+      read_samples_.splice(read_samples_.end(), not_read_samples_);
+
+      compute_ranks(sample_count, infos);
+
+      view_state_ = DDS::NOT_NEW_VIEW_STATE;
+    }
+
+    void take(SampleSequence& samples, InternalSampleInfoSequence& infos)
+    {
+      size_t sample_count = 0;
+
+      for (typename SampleList::const_iterator pos = read_samples_.begin(), limit = read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+      read_samples_.clear();
+
+      for (typename SampleList::const_iterator pos = not_read_samples_.begin(), limit = not_read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::NOT_READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+      not_read_samples_.clear();
 
-    PublicationSet publication_set;
+      compute_ranks(sample_count, infos);
 
-    DDS::ViewStateKind view_state;
-    DDS::InstanceStateKind instance_state;
+      view_state_ = DDS::NOT_NEW_VIEW_STATE;
+    }
 
-    bool unregister_instance(const T& sample, InternalEntity_wrch publication_handle)
+    void write(InternalEntity_wrch publication_handle,
+               const T& sample,
+               const DDS::DataReaderQos& qos)
     {
-      if (publication_set.erase(publication_handle)) {
-        samples.push_back(sample);
-        infos.push_back(InternalSampleInfo(ISIK_UNREGISTER, publication_handle));
+      publication_set_.insert(publication_handle);
+
+      if (view_state_ == DDS::NOT_NEW_VIEW_STATE && instance_state_ != DDS::ALIVE_INSTANCE_STATE) {
+        view_state_ = DDS::NEW_VIEW_STATE;
+      }
+
+      switch (instance_state_) {
+      case DDS::ALIVE_INSTANCE_STATE:
+        break;
+      case DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE:
+        ++disposed_generation_count_;
+        break;
+      case DDS::NOT_ALIVE_NO_WRITERS_INSTANCE_STATE:
+        ++no_writers_generation_count_;
+        break;
+      }
+
+      instance_state_ = DDS::ALIVE_INSTANCE_STATE;
 
-        if (publication_set.empty() && instance_state != DDS::NOT_ALIVE_DISPOSED_INSTANCE_STATE) {
-          instance_state = DDS::NOT_ALIVE_NO_WRITERS_INSTANCE_STATE;
+      if (qos.history.kind == DDS::KEEP_LAST_HISTORY_QOS) {
+        while (read_samples_.size() + not_read_samples_.size() >= qos.history.depth) {
```

Comment:
More warnings:



comparison of integer expressions of different signedness: ‘std::__cxx11::list<OpenDDS::DCPS::InternalDataReader<OpenDDS::DCPS::NetworkInterfaceAddress>::Instance::SampleHolder, std::allocator<OpenDDS::DCPS::InternalDataReader<OpenDDS::DCPS::NetworkInterfaceAddress>::Instance::SampleHolder> >::size_type’ {aka ‘long unsigned int’} and ‘const Long’ {aka ‘const int’} [-Wsign-compare]